cdk/shared_components/lambdas/run_athena_query_and_report_status.py

```diff
@@ -0,0 +1,116 @@
+"""
+A generic Lambda that can run an Athena query and wait for it to complete
+
+"""
+
+import time
+
+import boto3
+
+athena_client = boto3.client("athena")
+
+
+def get_named_query_by_name(query_name: str) -> dict:
+    """
+    Athena stores UUIDs and names for queries.
+
+    There is no way in the API to get a query by a name. `get_named_query`
+    takes a UUID, not a query name, annoyingly.
+
+    So, we need to get all the saved queries and return the first one
+    where the name matches.
+
+    """
+
+    # Get a list of all named query IDs in the workgroup
+    response = athena_client.list_named_queries(WorkGroup="dc-data-baker")
+    named_query_ids = response.get("NamedQueryIds", [])
+
+    # Iterate over each ID and retrieve the details
+    for query_id in named_query_ids:
+        query_details = athena_client.get_named_query(
+            NamedQueryId=query_id,
+        )
+        named_query = query_details.get("NamedQuery", {})
+        if named_query.get("Name") == query_name:
+            return named_query
+
+    # Raise if no query found
+    raise ValueError(f"Query {query_name} not found")
+
+
+def handler(event, context):
+    """
+    Supports both starting and then checking an Athena query.
+
+    Everything is configured by keys in the `event` dict.
+
+    `QueryName` is the named query to run. The query gets the event dict
+                as a template context
+
+    `QueryString` if this is passed in, `QueryName` is ignored. Designed
+                  to allow running ad-hox queries rather than saved queries.
+
+    `context`: this is passed to the query and anything here can be used
+               with `$foo` template substitution.
```

Comment:
is `event['context']` the same as `context`?

Reply:
No: confusing duplication of names here. `context` is a AWS Lambda standard for the context of the invocation. It's like `request.META` or whatever. `event["context"]` is the context that _we_ pass into the event.